main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -183,17 +181,24 @@ object Typer extends Phase[KindedAst.Root, TypedAst.Root] {
     *
     * Returns [[Err]] if a definition fails to type check.
     */
-  private def visitDefs(root: KindedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[Map[Symbol.DefnSym, TypedAst.Def], TypeError] = {
-    // Compute the results in parallel.
-    val results = ParOps.parMap(root.defs.values, visitDefn(_, Nil, root, classEnv))
-
-    // Sequence the results.
-    Validation.sequence(results) map {
-      case xs => xs.foldLeft(Map.empty[Symbol.DefnSym, TypedAst.Def]) {
-        case (acc, defn) => acc + (defn.sym -> defn)
+  private def visitDefs(root: KindedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext], oldRoot: TypedAst.Root, changeSet: ChangeSet)(implicit flix: Flix): Validation[Map[Symbol.DefnSym, TypedAst.Def], TypeError] =
+    flix.subphase("Defs") {
+      // Compute the stale and fresh definitions.
+      val (staleDefs, freshDefs) = changeSet.partition(root.defs, oldRoot.defs)
+
+      // println(s"Stale = ${staleDefs.keySet}")
+      // println(s"Fresh = ${freshDefs.keySet.size}")
+
+      // Process the stale defs in parallel.
+      val results = ParOps.parMap(staleDefs.values, visitDefn(_, Nil, root, classEnv))
```

Comment:
What about dependencies among defs?

Source 1:
```diff
- def f(): Int = 123
+ def f(): Bool = true
```

Source 2:
```diff
def g(): Int = f()
```

From what I can tell, this wouldn't re-check def `g`, because it's not in the change set. But `g` should now be a type error.

Reply:
Hmm. Yes. So then we have to build a dependency graph etc. :/